operators/pkg/controller/elasticsearch/driver/default.go

```diff
@@ -568,6 +512,161 @@ func removePodFromList(pods []corev1.Pod, pod corev1.Pod) []corev1.Pod {
 	return pods
 }
 
+func (d *defaultDriver) reconcileNodeSpecs(
+	results *reconciler.Results,
+	es v1alpha1.Elasticsearch,
+	podSpecBuilder esversion.PodTemplateSpecBuilder,
+	esClient esclient.Client,
+	observedState observer.State,
+) error {
+
+	actualStatefulSets, err := sset.RetrieveActualStatefulSets(d.Client, k8s.ExtractNamespacedName(&es))
+	if err != nil {
+		return err
+	}
+
+	nodeSpecResources, err := nodespec.BuildExpectedResources(es, podSpecBuilder)
+	if err != nil {
+		return err
+	}
+
+	// TODO: handle zen2 initial master nodes more cleanly
+	//  should be empty once cluster is bootstraped
+	var initialMasters []string
+	// TODO: refactor/move
+	for _, res := range nodeSpecResources {
+		cfg, err := res.Config.Unpack()
+		if err != nil {
+			return err
+		}
+		if cfg.Node.Master {
+			for i := 0; i < int(*res.StatefulSet.Spec.Replicas); i++ {
+				initialMasters = append(initialMasters, fmt.Sprintf("%s-%d", res.StatefulSet.Name, i))
+			}
+		}
+	}
+	for i := range nodeSpecResources {
+		if err := nodeSpecResources[i].Config.SetStrings(settings.ClusterInitialMasterNodes, initialMasters...); err != nil {
+			return err
+		}
+	}
+
+	// Phase 1: apply expected StatefulSets resources, but don't scale down.
+	// The goal is to:
+	// 1. scale sset up (eg. go from 3 to 5 replicas).
+	// 2. apply configuration changes on the sset resource, to be used for future pods creation/recreation,
+	//    but do not rotate pods yet.
+	// 3. do **not** apply replicas scale down, otherwise nodes would be deleted before
+	//    we handle a clean deletion.
+	for _, nodeSpec := range nodeSpecResources {
```

Comment:
Is this really a `nodeSpec`? The type would indicate either `nodeSpecResource` or `resource`?